cmd/ncproxy/ncproxy.go

```diff
@@ -0,0 +1,444 @@
+		caReq := &computeagent.AddNICInternalRequest{
+			ContainerID:  req.ContainerID,
+			NicID:        req.NicID,
+			EndpointName: req.EndpointName,
+		}
+		if _, err := client.AddNIC(ctx, caReq); err != nil {
+			return nil, err
+		}
+		return &ncproxygrpc.AddNICResponse{}, nil
+	}
+	return nil, status.Errorf(codes.FailedPrecondition, "No shim registered for namespace `%s`", req.ContainerID)
+}
+
+func (s *grpcService) DeleteNIC(ctx context.Context, req *ncproxygrpc.DeleteNICRequest) (*ncproxygrpc.DeleteNICResponse, error) {
+	log.G(ctx).WithFields(logrus.Fields{
+		"containerID":  req.ContainerID,
+		"nicID":        req.NicID,
+		"endpointName": req.EndpointName,
+	}).Info("DeleteNIC request")
+
+	if req.ContainerID == "" || req.EndpointName == "" || req.NicID == "" {
+		return nil, status.Error(codes.InvalidArgument, "received empty field in request")
+	}
+	if client, ok := containerIDToShim[req.ContainerID]; ok {
+		caReq := &computeagent.DeleteNICInternalRequest{
+			ContainerID:  req.ContainerID,
+			NicID:        req.NicID,
+			EndpointName: req.EndpointName,
+		}
+		if _, err := client.DeleteNIC(ctx, caReq); err != nil {
+			return nil, err
+		}
+		return &ncproxygrpc.DeleteNICResponse{}, nil
+	}
+	return nil, status.Errorf(codes.FailedPrecondition, "No shim registered for namespace `%s`", req.ContainerID)
+}
+
+//
+// HNS Methods
+//
+func (s *grpcService) CreateNetwork(ctx context.Context, req *ncproxygrpc.CreateNetworkRequest) (*ncproxygrpc.CreateNetworkResponse, error) {
+	log.G(ctx).WithFields(logrus.Fields{
+		"networkName": req.Name,
+		"type":        req.Mode.String(),
+		"ipamType":    req.IpamType,
+	}).Info("CreateNetwork request")
+
+	if req.Name == "" || req.Mode.String() == "" || req.IpamType.String() == "" || req.SwitchName == "" {
+		return nil, status.Error(codes.InvalidArgument, "received empty field in request")
+	}
+
+	// Check if the network already exists, and if so return error.
+	_, err := hcn.GetNetworkByName(req.Name)
+	if err == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "network with name `%s` already exists", req.Name)
+	}
+
+	// Get the layer ID from the external switch. HNS will create a transparent network for
+	// any external switch that is created not through HNS so this is what we're
+	// searching for here. If the network exists, the vSwitch with this name exists.
+	extSwitch, err := hcn.GetNetworkByName(req.SwitchName)
+	if err != nil {
+		if _, ok := err.(hcn.NetworkNotFoundError); ok {
+			return nil, status.Errorf(codes.NotFound, "no network/switch with name `%s` found", req.SwitchName)
+		}
+		return nil, fmt.Errorf("failed to get network/switch with name `%s`: %s", req.SwitchName, err)
+	}
+
+	// Get layer ID and use this as the basis for what to layer the new network over.
+	if extSwitch.Health.Extra.LayeredOn == "" {
+		return nil, status.Errorf(codes.NotFound, "no layer ID found for network `%s` found", extSwitch.Health.Extra.LayeredOn)
+	}
```

Comment:
just to be carefful: is it possible any of these fields could be empty in the event of an error?

Reply:
In regards to the `Health` and `Extra` fields? None of them are pointers so we'll never dereference anything if thats the worry, if not let me know . A problem I just noticed is that this uses the LayeredOn ID in the error message which will always be an empty string if we made it in this block haha. It should be using the extSwitches ID